### PR TITLE
Added Policy & Binding to the Haunted Longsword

### LINKED ISSUE #86864: Haunted longswords are unintuitive and a moderation issue.
### Client Version:

515.1640

### Issue Summary:

Haunted longsword - cult reward for saccing a heretic, is an almost immortal blade that can not be contained (since it can brake walls). This may be seen as a worthy reward for saccing a heretic, but it isnt.

The blade is allowed to kill WHOEVER it wants, by code. Its strong enough to 3tap armored cultists and can brake cult buidlings, which it will do, since you just ruined their heretic round.

Cult should not be punished for doing cult things - completing their objectives and saccing people. Heretics right now SHOULD never be sacced because of this blade, it can easily kill cult in minutes and is allowed to.

It should not be allowed to harm cultists or even be able to, since this is very unintuitive, and leads to cult getting punished for doing something good.

`tochat(trapped​entity, span_userdanger("You've been sacrificed to the Enemy, and trapped inside a haunted blade! While you cannot escape, you may help the Cult, your current wielder, or even pester everyone with what few abilities you kept."))"`

Pester = kill half the cult.

### Round ID:

[238786](https://scrubby.melonmesa.com/round/238786)

### Test Merge Information:

- [[TM ONLY] Adds a couple of sanity checks to storage code and temporary logging for when they are inevitably hit](https://www.github.com/tgstation/tgstation/pull/86842)

### Reproduction Steps:

1. Sacrafice a heretic
2. Recieve a Haunted longsword
3. Watch it kill cultists and brake your base
4. Try to throw it away
5.It can jump out of hands and use spells
6. Watch cult loose

### PR DESCRIPTION
**Pull request by carlarctg**
## About The Pull Request

Closes #86864

Added a policy message and associated config i think to when a heretic is soultrapped, allowing headmins to ~~ruin my feature~~ decide how they're allowed to act.

Haunted longswords can be, and start, binded. Binded swords are weaker than usual, being basically just talking cultist longswords. You can, however, unbind them, which will heavily buff their stats, unlock their powers, allow you use of a heretic's path spell, and let them attempt to fling around and cast spells.

Reduced its throwing damage, turning it from a 3-hit-crit to a 4-hit-crit.

Binding and unbinding needs a bible, a ritual dagger, or an amber focus, depending on what role you are.

Added type_english_list() which recieves a list of types and returns 'an x, a y, and a z' strings.
## Why It's Good For The Game

> Added a policy message and associated config i think to when a heretic is soultrapped, allowing headmins to ~~ruin my feature~~ decide how they're allowed to act.

Policy messages are required for new antags and a good idea besides. This slipped under the radar because of being bundled with a separate PR and the muddiness of it being 'new', but it's here now.

> Haunted longswords can be, and start, binded. Binded swords are weaker than usual, being basically just talking cultist longswords. You can, however, unbind them, which will heavily buff their stats, unlock their powers, allow you use of a heretic's path spell, and let them attempt to fling around and cast spells.

There's been some problems caused by an unfortunate lack of visibility on how the new sword is not necessarily your ally and may even immediately attempt retaliation, for, you know, being killed and sacrificed. To fix this, I've added a binding feature to it. The sword starts bound, but is fairly underwhelming in this state. It's suggested to talk to its wielders and tempt them into unbinding it, done by an UI button, which will allow it to use its spells and possibly backstab them, or help them in battle.

> Reduced its throwing damage, turning it from a 3-hit-crit to a 4-hit-crit.

The attacks were dealing too much damage, surprising hapless people being hit by random flying swords, who didn't react until it was too late.

This feature is still fucking awesome, and I spent some time wracking my brain on how to fix the given issues, and came up with this system which mantains and enhances the 'forbidden cursed sword' trope aspects of it while also stopping it from being as much of a noob trap as it used to be, Neutral antagonists that can help any side are cool and fun. 
## Changelog
:cl:
config: Added a policy message and associated config i think to when a heretic is soultrapped, allowing headmins to decide how they're allowed to act.
add: Haunted longswords can be, and start, binded. Binded swords are weaker than usual, being basically just talking cultist longswords. You can, however, unbind them, which will heavily buff their stats, unlock their powers, allow you use of a heretic's path spell, and let them attempt to fling around and cast spells. (it's a 4-tap on melee when bound, 3-tap if unbound)
add: Reduced its throwing damage, turning it from a 3-hit-crit to a 4-hit-crit. 
add: void haunted swords have void prison instead of void chill
code: Added type_english_list() which recieves a list of types and returns 'an x, a y, and a z' strings.
/:cl:

### CODE REVIEW

**Comment by Metekillot** [comment]
I'd keep its actual combat stats (thrown, melee) the same for either state, but besides that it looks pretty good. Lowering every single one of its stats for keeping it bound seems like a needless penalty; unbinding it is already pretty significant with the added spells and the option of having one more guy in the fight on your side -- your side, right?

Also I think non cultists who try to use it should get horrifically mutilated, and heretics can shatter it for knowledge points.

**Comment by Metekillot** [comment]
Also dude you should really state in the PR body that this doesn't just affect cultists who get a hold of the sword

**Comment by carlarctg** [comment]
> Also dude you should really state in the PR body that this doesn't just affect cultists who get a hold of the sword

?

**Comment by MoonDeer728** [comment]
void heretic have one spell and do slip his cultist on it

File: code/modules/antagonists/cult/sword_fling.dm [hunk 1]
```diff
@@ -16,7 +16,7 @@
 
 	cast_range = 6
 	active_msg = "You ready yourself to attempt to leap!"
-	var/obj/item/flinged_sword
+	var/obj/item/melee/cultblade/haunted/flinged_sword
```
Comment (Ben10Omintrix) [inline code/modules/antagonists/cult/sword_fling.dm:19]: what happens to this ability in the case the sword gets deleted? u might end up with a hanging ref there

**Comment by carlarctg** [comment]
todo make unbinding free and binding need the inplements

File: code/modules/antagonists/cult/cult_items.dm [hunk 2]
```diff
@@ -190,6 +194,107 @@ Striking a noncultist, however, will tear their flesh."}
...
+	if(istype(blade))
+		button_icon_state = "shade_[blade.bound ? "sealed" : "unsealed"]"
+		name = "[blade.bound ? "Unseal" : "Seal"] Spirit"
+
+	return ..()
+
+/obj/item/melee/cultblade/haunted/ui_action_click(mob/living/user, actiontype)
+	if(DOING_INTERACTION_WITH_TARGET(user, src))
+		return // gtfo
+	// if(IS_NORMIE())
```
Comment (MrMelbert) [inline code/modules/antagonists/cult/cult_items.dm:226]: What is this line for? 

Are normal crew members supposed to be able to bind / unbind it? Or only magically inclined ones?

File: code/modules/antagonists/cult/cult_items.dm [hunk 3]
```diff
@@ -252,28 +358,64 @@ Striking a noncultist, however, will tear their flesh."}
...
+			LAZYADD(path_sword_actions, instanced_spell)
+			instanced_spell.overlay_icon_state = "bg_cult_border" // for flavor, and also helps distinguish
 
 	if(wielder_spells)
 		for(var/datum/action/cooldown/spell/wielder_spell as anything in wielder_spells)
-			path_wielder_action = new wielder_spell(src)
-			wielder_spell?.overlay_icon_state = "bg_cult_border"
+			var/datum/action/cooldown/spell/instanced_spell = new wielder_spell(trapped_entity)
+			LAZYADD(path_wielder_actions, new wielder_spell(src))
+			instanced_spell.overlay_icon_state = "bg_cult_border"
```
Comment (MrMelbert) [inline code/modules/antagonists/cult/cult_items.dm:374]: Two spells are created here, and one doesn't get a cult background, is that intentional?

File: code/modules/antagonists/cult/cult_items.dm [hunk 4]
```diff
@@ -252,28 +358,64 @@ Striking a noncultist, however, will tear their flesh."}
...
 
 /obj/item/melee/cultblade/haunted/equipped(mob/user, slot, initial)
 	. = ..()
-	if(slot & ITEM_SLOT_HANDS)
-		path_wielder_action?.Grant(user)
+	if((!(slot & ITEM_SLOT_HANDS)) || bound)
+		return
+	for(var/datum/action/cooldown/spell/wielder_spell as anything in path_wielder_actions)
+		wielder_spell?.Grant(user)
```
Comment (MrMelbert) [inline code/modules/antagonists/cult/cult_items.dm:381]: If you are expecting the list to hold nulls
```suggestion
	for(var/datum/action/cooldown/spell/wielder_spell in path_wielder_actions)
		wielder_spell.Grant(user)
```
If not
```suggestion
	for(var/datum/action/cooldown/spell/wielder_spell as anything in path_wielder_actions)
		wielder_spell.Grant(user)
```

File: code/modules/antagonists/cult/cult_items.dm [hunk 5]
```diff
@@ -252,28 +358,64 @@ Striking a noncultist, however, will tear their flesh."}
...
+	binding_filters_update()
 
 /obj/item/melee/cultblade/haunted/dropped(mob/user, silent)
 	. = ..()
-	path_wielder_action?.Remove(user)
+	for(var/datum/action/cooldown/spell/wielder_spell as anything in path_wielder_actions)
+		wielder_spell?.Remove(user)
```
Comment (MrMelbert) [inline code/modules/antagonists/cult/cult_items.dm:387]: Same here
```suggestion
	for(var/datum/action/cooldown/spell/wielder_spell in path_wielder_actions)
		wielder_spell.Remove(user)
```
```suggestion
	for(var/datum/action/cooldown/spell/wielder_spell as anything in path_wielder_actions)
		wielder_spell.Remove(user)
```

**Review by MrMelbert — CHANGES_REQUESTED** [review]

File: code/modules/antagonists/cult/cult_items.dm (hunk 2 above)
Comment (MrMelbert) [reply code/modules/antagonists/cult/cult_items.dm:226]: (IMO, normies shouldn't be able to un/bind it. But you should also update the examine message so it doesn't show up for normies as well)

File: code/modules/antagonists/cult/sword_fling.dm (hunk 1 above)
Comment (carlarctg) [reply code/modules/antagonists/cult/sword_fling.dm:19]: it nulls the reference on destruction either way

File: code/modules/antagonists/cult/cult_items.dm (hunk 2 above)
Comment (carlarctg) [reply code/modules/antagonists/cult/cult_items.dm:226]: i think it'd be pretty lame if 90% of the crew were unable to interact with the feature

Comment (carlarctg) [reply code/modules/antagonists/cult/cult_items.dm:226]: besides the trope of unsealing an ancient horror works better if any dumb fool (assistant) can be sweet talked into it

File: code/modules/antagonists/cult/cult_items.dm (hunk 3 above)
Comment (carlarctg) [reply code/modules/antagonists/cult/cult_items.dm:374]: oop

**Comment by carlarctg** [comment]
COMMITTING TOMROROW NAP TIME

**Comment by carlarctg** [comment]
why did it just now add that

**Review by MrMelbert — APPROVED** [review]

**Pull request merged**